File: c_files/src/Person.c
/**
*
*@author Sude Andurman	sude.andurman@ogr.sakarya.edu.tr
*@since 03.05.2025
*<p>
*	Bilgisayar Mühendisliği 2. Sınıf 
*	Programlama Dilleri Prensipleri 1/C şubesi 
*</p>
*/

#include "Person.h"

/* PRIVATE FUNSTIONS */
void	reduceHoursLeftToLive(Person this)
{
	/* CHECKING PARAMETERS AND VARIABLES THAT WE WILL USE IN THIS FUNCTION */
    if (this == NULL || 
        this->currentVehicle == NULL || 
        this->currentVehicle->state == NULL ||
        this->currentVehicle->currentPlanet == NULL ||
        this->currentVehicle->destPlanet == NULL ||
        this->hoursLeftToLive <= 0)
        return;
    
	/* REDUCE HOURS LEFT TO LIVE */
    if (this->hoursLeftToLive > 0)
    {
        if (strcmp(this->currentVehicle->state, "Bekliyor") == 0)
        {
            this->hoursLeftToLive -= this->currentVehicle->currentPlanet->getDecayRate(this->currentVehicle->currentPlanet);
        }
        else if (strcmp(this->currentVehicle->state, "Vardi") == 0)
        {
            this->hoursLeftToLive -= this->currentVehicle->destPlanet->getDecayRate(this->currentVehicle->destPlanet);
        }
        else
        {
            this->hoursLeftToLive -= 1; // when the state is "Yolda"
        }
    }

}
/* END OF PRIVATE FUNCTIONS */

Person	new_Person(const char* name, int age, double hoursLeftToLive, const char* currentVehicleName)
{
	/* CHECKING THE PARAMETERS */
	if (name == NULL || currentVehicleName == NULL)
	{
		printf("Memory allocation failed at new_Person\n");
		return (NULL);
	}

	/* MEMORY ALLOCATION FOR PERSON */
	Person this = (Person)malloc(sizeof(struct PERSON));
	if (this == NULL)
	{
		printf("Memory allocation failed at new_Person\n");
		return (NULL);
	}

	/* MEMORY ALLOCARION FOR ITIMELISTENER */
	this->listener = new_ITimeListener();
	if (this->listener == NULL)
	{
		printf("Memory allocation failed for listener!\n");
		free(this);
		return NULL;
	}

	/* SET INITAL VALUES */
	this->name = strdup(name);
	this->age = age;
	this->currentVehicleName = strdup(currentVehicleName);
	this->hoursLeftToLive = hoursLeftToLive;

	/* LINKS */
	this->setInitials_Person = &setInitials_Person;
	this->isAlive = &isAlive;
	this->delete_Person = &delete_Person;
	this->listener->onTimePass = &onTimePass_Person; // interface implementation method
    /* END OF LINKS */
	
	return (this);
}

void	setInitials_Person(const Person this, ArrayList spaceShipList)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL ||
		spaceShipList == NULL ||
		this->currentVehicleName == NULL)
	{
		printf("EROR: empty parameter at setInitials_Person!\n");
		return;
	}

	this->currentVehicle = NULL; // initial value for now
	for (size_t i = 0; i < spaceShipList->size ; i++)
	{
		SpaceShip spaceShip = (SpaceShip)(spaceShipList->getByIndex(spaceShipList, i));
		if (spaceShip != NULL && spaceShip->name != NULL)
		{
			/* IF SHIP EXISTS SET VALUE */
			if (equalsIgnoreCase(this->currentVehicleName, spaceShip->name))
			{
				this->currentVehicle = spaceShip;
				break;
			}
		}
	}
}

bool	isAlive(const Person this)
{
	/* CHECKING THE PARAMETERS */
    if (this == NULL)
    {
        printf("Eror: person is null at isAlive \n");
        return false;
    }

	/* ISALIVE */
    if (this->hoursLeftToLive <= 0.0)
    {
        return false;
    }else
        return true;
}

void	delete_Person(Person this)
{
    if (this == NULL)
        return;
    
    if (this->name != NULL)
        free(this->name);

    if (this->currentVehicleName != NULL)
        free(this->currentVehicleName);

    if (this->listener != NULL)
        free(this->listener);

    free(this);
}

/* OVERRIDE METHOD */
void	onTimePass_Person(const Person this)
{
    if (this == NULL)
    {
        printf("Error: onTimePass_Person 'this' NULL!\n");
        return;
    }
    reduceHoursLeftToLive(this);
}

File: java_files/pdp_planet_travel_sim/src/app/Person.java
/**
*
*@author Sude Andurman	sude.andurman@ogr.sakarya.edu.tr
*@since 03.04.2025
*<p>
*	Bilgisayar Mühendisliği 2. Sınıf 
*	Programlama Dilleri Prensipleri 1/C şubesi 
*</p>
*/


package app;

/* IMPORTS */
import java.util.List;
/* END OF IMPORTS */

public class Person implements ITimeListener
{
	private String		name;			// TO SET THE NAME THAT WE GET FROM THE FILE
	private String		currentVehicle;	// TO SET THE CURRENT VEHICLE THAT WE GET FROM THE FILE
	private int			age;			// TO SET THE AGE THAT WE GET FROM THE FILE
	private int			hoursLeftToLive;// TO SET THE LIFETIME THAT WE GET FROM THE FILE

	/* CONSTRUCTERS */
	
	public Person(String name, int age, int hoursLeftToLive ,String currentVehicle)
	{
		this.setName(name);
		this.setAge(age);
		this.setHoursLeftToLive(hoursLeftToLive);
		this.setCurrentVehicle(currentVehicle);
		Time.addListener(this);
	}
	
	/* END OF CONSTRUCTERS */
	
	/* SETTER METHODS */
	
	private void	setName(String name)
	{
		this.name = name;
	}
	private void	setCurrentVehicle(String currentVehicle)
	{
		this.currentVehicle = currentVehicle;
	}
	private void	setAge (int age)
	{
		this.age = age;
	}
	private void	setHoursLeftToLive(int hours)
	{
		if (hours < 0) // TO PREVENT FROM ERORS SET THE LIFETIME "0"
		{
			hoursLeftToLive = 0;
		}
		else
		{
			this.hoursLeftToLive = hours;
		}	
	}
	
	/* END OF SETTER METHODS */
	
	/* GETTER METHODS */
	
	public String	getCurrentVehicle()
	{
		return (currentVehicle);
	}
	public int		getHoursLeftToLive()
	{
		return (hoursLeftToLive);
	}
	public String	getName()
	{
		return (name);
	}
	public int		getAge()
	{
		return (age);
	}
	/* END OF GETTER METHODS */
	
	/* METHODS */
	
	private void	reduceHoursLeftToLive()
	{
		if (isAlive())
		{
			hoursLeftToLive--;
		}
	}
	
	public boolean	isAlive()
	{
		if (hoursLeftToLive == 0)
		{
			// IF THE PERSON IS NOT ALIVE NO NEED TO BE RELATED WITH THE TIME ANYMORE
			Time.removeListener(this);
			return (false);
		}
		else
			return (true);
	}
	
	public boolean	isValid(List<SpaceShip> ships)// TO FIND THE SHIP THAT PERSON CURRENTLY IS IN
	{
		// IF THERE ANY MATCHED PLANET NAME RETURN TRUE TO ADD THE PERSON UPDATED LIST
		return (ships.stream()
	           .anyMatch(ship -> ship.getName().equalsIgnoreCase(this.currentVehicle)));
		
	}
	
	@Override
	public void		onTimePass()
	{
		// TRIGGER THE FUNCTIONS THAT RELATED TO THE TIME
		reduceHoursLeftToLive();

	}
}
